enclave-manager/web/src/components/FileTree.tsx

```diff
@@ -0,0 +1,123 @@
+import { Button, Flex, Text } from "@chakra-ui/react";
+import React, { useCallback, useMemo, useState } from "react";
+import { AiFillFile, AiFillFolder, AiFillFolderOpen } from "react-icons/ai";
+import { isDefined } from "../utils";
+import { FileSize } from "./FileSize";
+
+/**
+ * This file tree component recursively renders itself to present a file tree.
+ * To keep this performant the nodes (DirectoryNode and FileNode) must make use of
+ * useCallback and useMemo. This allows the React.memo around FileTreeMode to function
+ * and skip rendering unchanged components.
+ */
+
+export type FileTreeNode = {
+  name: string;
+  size?: bigint;
+  childNodes?: FileTreeNode[];
+};
+
+type FileTreeProps = {
+  nodes: FileTreeNode[];
+  selectedFilePath?: string[];
+  onFileSelected: (selectedFilePath: string[]) => void;
+  // Internal prop used for padding
+  _isChildNode?: boolean;
+};
+
+export const FileTree = ({ nodes, selectedFilePath, onFileSelected, _isChildNode }: FileTreeProps) => {
+  return (
+    <Flex flexDirection={"column"} pl={_isChildNode ? "22px" : undefined} w={"100%"}>
+      {nodes.map((node, i) => (
+        <FileTreeNodeComponent
+          key={node.name}
+          node={node}
+          selectedFilePath={
+            isDefined(selectedFilePath) && selectedFilePath.length > 0 && selectedFilePath[0] === node.name
+              ? selectedFilePath
+              : undefined
+          }
+          onFileSelected={onFileSelected}
+        />
+      ))}
+    </Flex>
+  );
+};
+
+type FileTreeNodeComponentProps = {
+  node: FileTreeNode;
```

Comment:
Would making this optional/`?` alleviate some of the checking and typecasting in the `FileTreeNodeComponent` and `DirectoryNode`? If not, would love to learn why this pattern is better than having an optional member of the `FileTreeNodeComponentProps`

Reply:
Ah thanks for asking. Here I've chosen to keep `node` mandatory because this component is rendering a `FileTreeNode` - specifically it's deciding whether to render a directory node or a file node. It is never the case that the `node` is optional.

I agree the casting in `FileTreeNodeComponent` is not ideal - it's needed because typescript is not able to infer parent types (ie the type of `props`) based on the type of child types (ie the type of `childNodes` tells us the type of `node`). https://github.com/microsoft/TypeScript/issues/42384. An alternative would be rather than using our generic `isDefined` typeguard (which is a generally useful utility) we could implement a typeguard specificly to assert `FileTreeNodeComponentProps & { node: { childNodes: FileTreeNode[] } }`